audit/audit.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package audit provides an audit log writer for access to secrets.
+package audit
+
+import (
+	"encoding/json"
+	"io"
+	"net/netip"
+	"os"
+
+	"github.com/tailscale/setec/acl"
+	"github.com/tailscale/setec/types/api"
+)
+
+// Principal is the identity of a client taking action on the secrets
+// service.
+type Principal struct {
+	Hostname string     `json:"hostname"`
+	IP       netip.Addr `json:"ip"`
+	User     string     `json:"user,omitempty"`
+	Tags     []string   `json:"tags,omitempty"`
+}
+
+// Entry is an audit log entry.
+type Entry struct {
+	// Principal is the client who is doing something.
+	Principal Principal `json:"principal"`
+	// Action is the action being performed on a secret.
+	Action acl.Action `json:"action"`
+	// Secret is the name of the secret being acted upon.
+	Secret string `json:"secret"`
+	// SecretVersion is the version of the secret being acted upon, if
+	// applicable, or api.SecretVersionDefault if a version doesn't
+	// make sense for the action (e.g. listing a secret).
+	SecretVersion api.SecretVersion `json:"secretVersion,omitempty"`
+}
+
+// Writer is an audit log writer.
+type Writer struct {
+	w io.Writer
+}
+
+// New returns a Writer that outputs audit log entries to w. If w also
+// implements io.Closer, Writer.Close closes w. If w also implements a
+// Sync method with the same signature as os.File, Writer.Sync calls
+// w.Sync.
+func New(w io.Writer) *Writer {
+	return &Writer{w}
+}
+
+// NewFile returns a Writer that outputs audit log entries to a file
+// at path, creating it if necessary.
+func NewFile(path string) (*Writer, error) {
+	f, err := os.OpenFile(path, os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0600)
+	if err != nil {
+		return nil, err
+	}
+	return New(f), nil
+}
+
+// Sync commits the current contents of the file to stable storage if
+// the Writer was created with a sink that itself implements Sync, or
+// else does nothing successfully.
+func (l *Writer) Sync() error {
+	if s, ok := l.w.(syncer); ok {
+		return s.Sync()
+	}
+	return nil
+}
+
+// Close closes the Writer if the writer was created with a sink that
+// implements io.Closer, or else does nothing successfully.
+func (l *Writer) Close() error {
+	if err := l.Sync(); err != nil {
+		return err
+	}
+	if c, ok := l.w.(io.Closer); ok {
+		return c.Close()
+	}
+	return nil
+}
+
+type syncer interface {
+	Sync() error
+}
+
+// WriteEntry writes e to the audit log.
+func (l *Writer) WriteEntry(e *Entry) error {
+	bs, err := json.Marshal(e)
```

Comment:
If we carry a `json.Encoder` in the struct, its `Encode` method will add the newline for us. (We'd still want the writer, so we can flush and close and whatnot, but the encoder doesn't buffer so it's fine to have both)

Reply:
TIL, done.

Reply:
The downside is the encoder benchmarks slower than json.Marshal for some reason. Not enough to worry about for what we're doing here, though, methinks.